terraform/eval_for_each.go

```diff
@@ -0,0 +1,65 @@
+package terraform
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/hcl2/hcl"
+	"github.com/hashicorp/terraform/tfdiags"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// evaluateResourceForEachExpression interprets a "for_each" argument on a resource.
+//
+// Returns a cty.Value map, and diagnostics if necessary. It will return nil if
+// the expression is nil, and is used to distinguish between an unset for_each and an
+// empty map
+func evaluateResourceForEachExpression(expr hcl.Expression, ctx EvalContext) (forEach map[string]cty.Value, diags tfdiags.Diagnostics) {
+	if expr == nil {
+		return nil, nil
+	}
+
+	forEachVal, forEachDiags := ctx.EvaluateExpr(expr, cty.DynamicPseudoType, nil)
+	diags = diags.Append(forEachDiags)
+	if diags.HasErrors() {
+		return nil, diags
+	}
+
+	if forEachVal.IsNull() {
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Invalid for_each argument",
+			Detail:   `The given "for_each" argument value is unsuitable: the given "for_each" argument value is null. A map, or set of strings is allowed.`,
+			Subject:  expr.Range().Ptr(),
+		})
+		return nil, diags
+	}
+
+	if !forEachVal.CanIterateElements() || forEachVal.Type().IsListType() {
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Invalid for_each argument",
+			Detail:   fmt.Sprintf(`The given "for_each" argument value is unsuitable: the "for_each" argument must be a map, or set of strings, and you have provided a value of type %s.`, forEachVal.Type().FriendlyName()),
+			Subject:  expr.Range().Ptr(),
+		})
+		return nil, diags
+	}
+
+	if forEachVal.Type().IsSetType() {
+		if forEachVal.Type().ElementType() != cty.String {
+			diags = diags.Append(&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Invalid for_each set argument",
+				Detail:   fmt.Sprintf(`The given "for_each" argument value is unsuitable: "for_each" supports maps and sets of strings, but you have provided a set containing type %s.`, forEachVal.Type().ElementType().FriendlyName()),
+				Subject:  expr.Range().Ptr(),
+			})
+			return nil, diags
+		}
+	}
+
+	// If the map is empty ({}), return an empty map, because cty will return nil when representing {} AsValueMap
+	if forEachVal.LengthInt() == 0 {
```

Comment:
If you change this to 
```
if !forEachVal.IsKnown() || forEachVal.LengthInt() == 0 {
```
then it fixes the panic I saw. I don't completely understand what is going on, but I think it might evaluate it first without the injected variables, and then later evaluates it with the variables. (Maybe it is better to have a seperate branch to handle the unknown case?)